In [1]:
# Diego Benavides
#feb 23 2022
#Midterm assignment 

## The Grand Tour  

### The year is 2340. Human kind has colonized the whole Solar system. Travelling between planets, moons and minor bodies with one's spaceship is as easy as driving a car between cities used to be in the 21st century. There are still, however, some extra considerations to take into account as all these bodies evolve around the Sun and one needs to organize their trip carefully, or risk wasting *a lot* of fuel. Your mission is to create a code that will enable navigation through the Solar system and let the user know when is best to travel to their destination.

<img src="grand_tour_nasa.png" width=300 height=500 />
Credit: NASA

## 1. The setup (10 points total)

At this point you should have made a folder named Midterm_yourname and have moved this notebook in there. (**2 points**)

Don't forget that a good log is part of a successful mission! Start your log please! (**3 points**)

You will need to import the necessary package(s) to do numerical problems. Start by doing all your imports here. (**5 points**)

In [1]:
import numpy as np

In [2]:
import matplotlib.pyplot as plt 

## 2. Well begun is half done. (30 points total + 2 bonus)

### We need to ask the user of our code where they travel from/to.

A) Make a variable named *start_location* that will hold the information of where the 
traveler starts from. Make a variable *destination_location* that will hold the information 
of where the traveler wants to end. Your code should ask for input from the user about 
where they want to travel to/from and store the input in these variables. (**5 points**) 

In [3]:
print('Type in your starting location and then type in your desired destination location:\n')

start_location = str(input())

destination_location = str(input())


Type in your starting location and then type in your desired destination location:

Earth
Mars


### Now we need to make sure that the traveler gave us valid input. 

B) Make a list named *planets_and_moons* that contains the names of all possible planets and moons the traveler can travel between. The list should contain only the following bodies: 

Mercury, Venus, Earth, Moon, Ceres, Mars, Jupiter, Io, Europa, Saturn, Titan, Neptune, Pluto, Charon. (**5 points**) 



In [4]:
planets_and_moons = ['Mercury' , 'Venus' , 'Earth' , 'Moon' , 'Ceres' , 'Mars' , 'Jupiter' , 'Io' , 'Europa' , 'Saturn' , 'Titan' , 'Neptune' , 'Pluto' , 'Charon']
#print(type(planets_and_moons))

C) Make a code that checks whether the input given from the traveler is available among the allowed locations a trip can be made to/from. (**2 points**) 

If both are valid inputs the code should print a message to let the traveler know the code will proceed (**3 point**).

If any of the inputs are not allowed, the code should let the traveler know that they need to give a new start or destination point, and inform them what options are available. The code should then ask for the corrected start or destination points. (**5 points**)

Note: at this point we will assume that the traveler entered a correct location and did not do a mistake a second time. **Bonus 2 points:** what would you need to add to the code to check that this assumption is correct, and let the traveler know they made a new mistake? Add functionality that could account for the user making repeated mistakes entering the locations until they entered it in correctly. Explain with a comment under the code you just wrote. 

In [5]:
if start_location in planets_and_moons:
    if destination_location in planets_and_moons:
        print(f'Both {start_location} and {destination_location} are valid locations. The Journey will proceed!')
    else:
        print(f'One or more of your locations is invalid.\nYour possible options are {planets_and_moons}.\nInput new locations:\n')
        start_location = str(input())
        destination_location = str(input())
else:
    print(f'One or more of your locations is invalid.\nYour possible options are {planets_and_moons}.\nInput new locations:\n')
    
    start_location = str(input())
    destination_location = str(input())
    

Both Earth and Mars are valid locations. The Journey will proceed!


In [ ]:
#In order to check if the new locations are correct or incorrect you would have to 
#repeat the code above for the new locations. But this could go on forever if you wanted to
#by continuing to check if the new locations are incorrect, and then do the same for the ones
#after that and after that.

### Test that your code works.  

D-1) Try a trip starting at Demos and destination Io. Run the code so far, and make a screenshot showing that the code warns you that the starting point is not valid. Save the screenshot with the appropriate name to your folder. (**5 points**)

D-2) Change the starting point to Jupiter. Make a screenshot showing that the code lets you know you will proceed with the trip. Save the screenshot with the appropriate name to your folder. (**5 points**)


In [ ]:
# Now I will take a screen shot for the above prompts D-1 and D-2.

## 3. The solar system nagivation. (20 points total)

Now that we know what the traveler wants to do, we need to let them know how long the trip will be and how much time it will take them. 

A) File solar_system_date_1.dat has information about the distances (in AU) between all bodies on a given date. If you open the file, you will see that it contains a 14 x 14 table (as we have 14 celestial bodies that we can travel between) with the distances between the bodies. The  Nth line and Nth column are a given body, so that the diagonal is zeros (as the distance from A to A is zero). The distance between Earth and the Moon, e.g., is 0.0030 AU so that line 3 (#4 of the file) column 2 of your data (i.e., the distance from the Earth to the Moon) is 0.0030 (note that the same is true for line 2 column 3 as this is the distance from the Moon to the Earth). 

Use the appropriate numpy function to read the data into variable *space_distances_date1*. (**3 points**)

In [6]:
space_distances_date_1 = np.loadtxt('solar_system_date_1.dat', comments = '#')

#print(space_distances_date1)

B) Next you will need to find which solar system bodies the traveler wants to travel from/to when they give you their navigation start and end points (like above Jupiter/Io). Use your preferred method to get the index of the bodies (name the indexes *start_body_i* and *destination_body_i*) from your planets_and_moons list. (**4 points**) 

In [7]:
start_body_i = planets_and_moons.index(start_location)

print(start_body_i)

2


In [8]:
destination_body_i = planets_and_moons.index(destination_location)

print(destination_body_i)

5


C) Using the space_distances_date1 data and indexes start_body_i and destination_body_i you just calculated, find the distance the traveler will travel. Print an informative statement to let them know what the distance between their start and end points are (e.g., 'You will travel 10AU from the Moon to Calisto' ) (**5 points**)

In [9]:
desired_travel_distance = space_distances_date_1[ destination_body_i , start_body_i ]

#print(desired_travel_distance)

In [10]:
print(f'The distance between {start_location} and {destination_location} is {desired_travel_distance} AU')

The distance between Earth and Mars is 2.113 AU


D) Assume that the traveler can travel at an average speed of 0.001AU/hr. Calculate the average travel time it will take them to go from their start to their destination body. Calculate how many Earth years (travel_time_years; assume 365 days in the year), Earth days (travel_time_days), Earth hours (travel_time_hours) and Earth minutes (travel_time_minutes) the trip will take. </br>
Present this result in the format: "(Earth_Years) years (Earth_Days) days, (Earth_Hours) hours (Earth_Mins) mins". With the corresponding values in the parentheses, do not include the parentheses in your final answer. Print an informative statement that lets the traveler know how long their trip will take. (**4 points**) 

In [11]:
average_speed = 0.001

In [12]:
travel_time_years = (desired_travel_distance/average_speed) / (24*365)

print(travel_time_years)

0.24121004566210047


In [13]:
travel_time_days =  (travel_time_years - int(travel_time_years)) * 365

print(travel_time_days)

88.04166666666667


In [14]:
travel_time_hours = (travel_time_days - int(travel_time_days)) * 24

print(travel_time_hours)

1.0000000000001137


In [15]:
travel_time_minutes = (travel_time_hours - int(travel_time_hours)) * 60

print(travel_time_minutes)

6.821210263296962e-12


In [16]:
print(f" Your trip will take {int(travel_time_years)} year(s), {int(travel_time_days)} day(s), {int(travel_time_hours)} hour(s), and {int(travel_time_minutes)} minute(s)")

 Your trip will take 0 year(s), 88 day(s), 1 hour(s), and 0 minute(s)


E) Turn the program you wrote so far into a function *planet_tour* that takes as input the names of the two bodies (start, destination)the traveler wants to travel between, and outputs the necessary information (total distance, travel time). Make sure you include an appropriate informative docstring! Save the function in a separate file. (**2 points**)

In [ ]:
# code for planet_tour() function:

#def planet_tour( start , destination ):
 #   """This function will take in the given starting location and the given destination location, and return the
  #  distance between the two locations and the time taken to travel between them moving at 0.001 AU/hr
   # INPUT: start location input at the beginning of the code, destination location input in the beginning 
    #OUTPUT: distance between the two inputs, travel time between the two places with a speed of 0.001 AU/hr"""
    
    #planets_and_moons = ['Mercury' , 'Venus' , 'Earth' , 'Moon' , 'Ceres' , 'Mars' , 'Jupiter' , 'Io' , 'Europa' , 'Saturn' , 'Titan' , 'Neptune' , 'Pluto' , 'Charon']
    #print('Type in your starting location and then type in your desired destination location:\n')
    
    #start = str(input())
    
    #destination = str(input())
    
    #start_location = start
    
    #destination_location = destination
    
    #if start_location in planets_and_moons:
     #   if destination_location in planets_and_moons:
      #      print(f'Both {start_location} and {destination_location} are valid locations. The Journey will proceed!\n')
    #else:
     #   print(f'One or more of your locations is invalid.\nYour possible options are {planets_and_moons}.\nInput new locations:\n')
      #  new_start_location = str(input())
       # new_destination_location = str(input())
    
    
    #space_distances_date1 = np.loadtxt('solar_system_date_1.dat', comments = '#')
    
    #start_body_i = planets_and_moons.index(start_location)
    #destination_body_i = planets_and_moons.index(destination_location)

    #desired_travel_distance = space_distances_date1[ destination_body_i , start_body_i ]
    
    #total_distance = f"Your total travel distance is {desired_travel_distance} AU"
    
    #average_speed = 0.001
    #travel_time_years = (desired_travel_distance/average_speed) / (24*365)
    #travel_time_days =  (travel_time_years - int(travel_time_years)) * 365
    #travel_time_hours = (travel_time_days - int(travel_time_days)) * 24
    #travel_time_minutes = (travel_time_hours - int(travel_time_hours)) * 60
    
    #travel_time = f" Your trip will take {int(travel_time_years)} year(s), {int(travel_time_days)} day(s), {int(travel_time_hours)} hour(s), and {int(travel_time_minutes)} minute(s)"
    
    #return (total_distance , travel_time )###


In [30]:
#Now I will save the function into a separate file titled midterm_funct_diegobenavidesucf.py by copy and pasting it in there
#and I will turn the cell above into text so the function doesnt go into my notebook's memory 
#and I can prove that I am actually importing the function into this notebook from a .py file.

In [18]:
#testing function
from midterm_funct_diegobenavidesucf import planet_tour

print(planet_tour('Earth' , 'Mars'))

Type in your starting location and then type in your desired destination location:

Demos
Io
One or more of your locations is invalid.
Your possible options are ['Mercury', 'Venus', 'Earth', 'Moon', 'Ceres', 'Mars', 'Jupiter', 'Io', 'Europa', 'Saturn', 'Titan', 'Neptune', 'Pluto', 'Charon'].
Input new locations:

Jupiter
Io
The best date to travel is Date 3 , the distance is 0.0018 AU.

Your trip will take 0 year(s), 0 day(s), 1 hour(s), and 47 minute(s)

Your journey will require -0.0 refueling stops.
No stops needed for your trip.

('The best date to travel is Date 3 , the distance is 0.0018 AU.', 'Your trip will take 0 year(s), 0 day(s), 1 hour(s), and 47 minute(s)')


F) Test that your program so far works. Import the function and call it for a trip from the Earth to Mars. Your trip should take 88 days and 1 hour. Note that this is *much* faster than what the trip takes nowadays, where we need the better part of a year (9 months or more) to get there. (**2 points**)


In [19]:
from midterm_funct_diegobenavidesucf import planet_tour

print(help(planet_tour))

Help on function planet_tour in module midterm_funct_diegobenavidesucf:

planet_tour(start, destination)
    This function will take in the given starting location and the given destination location, and return the
    distance between the two locations for the best case and the time taken to travel between them 
    moving at 0.001 AU/hr for the best case.
    The function also returns how many stops of refueling are required, and makes a .txt file to send to 
    the stations mentioning where the refueling stops are needed. It also prints the date for the best case with the 
    distance for the best case of each refuel stop with a 4 digit precision.
    
    INPUT: start location input at the beginning of the code, destination location input in the beginning 
    OUTPUT: distance between the two inputs for the best case, 
            and the travel time between the two places with a speed of 0.001 AU/hr for the best case

None


In [21]:
#planet_tour( 'Earth' , 'Mars' )

In [20]:
print(planet_tour( 'Earth' , 'Mars' ))

Type in your starting location and then type in your desired destination location:

Earth
Mars
Both Earth and Mars are valid locations. The Journey will proceed!

The best date to travel is Date 3 , the distance is 2.0230 AU.

Your trip will take 0 year(s), 84 day(s), 7 hour(s), and 0 minute(s)

Your journey will require 3.0 refueling stops.


My trip starts at Earth and ends at Mars. I will need to encounter a stop at 0.6500 AU.
I will need to encounter a stop at 1.3000 AU.
I will need to encounter a stop at 1.9500 AU.

('The best date to travel is Date 3 , the distance is 2.0230 AU.', 'Your trip will take 0 year(s), 84 day(s), 7 hour(s), and 0 minute(s)')


## 4. Refueling is still a thing. (25 points total)

Humans are able to travel through space easily, but refueling is still a thing as we have not discovered ways to use an infinite source of energy. It is important to let people know not only how long their trip will take, but also how much fuel they will need on their trip. The Solar system is full of refueling stations, but as they are in orbit we need to send them messages about where to meet our spaceship and how much fuel we will need. 

You will need to change the function you just made to include this information and print a statement that you will send the refueling stations.

A) Assuming your spaceship needs refueling every 0.65AU calculate how many refueling stops your trip will make. The code should print a message to let the user know how many stops they will need for refueling. Assume you start every trip with a full tank. Remember to round up when needed (1.2 stops doesn't make sense: 2 stops does). ( **5 points**)


In [21]:
#I will first make the refueling count calculation in this cell before adding it to the function.
#np.ceil() will round up my numbers so it makes sense
#But assuming you start with a full tank,
#you have to subtract one stop after using division to calculate how many stops are needed.

refuel_distance = 0.65

expected_stops = np.ceil(((desired_travel_distance / refuel_distance)-1))

print(f"Your journey will require {expected_stops} refueling stops")

Your journey will require 3.0 refueling stops


In [23]:
#I will now add this code to the function by copy and pasting into the midterm_funct_diegobenavidesucf.py file

In [23]:
#testing the function now:
#from midterm_funct_diegobenavidesucf import planet_tour
#print(planet_tour( 'Earth' , 'Mars'))

B) If you will need refueling, the code should write a file *message_to_stations.txt* that will be sent to the refueling stations with information about: your start and end point, where they will need to meet with you (e.g., 'My trip starts at Earth and ends at Mars. I will need to encounter stations for a refueling at 0.65AU, 1.3AU and 1.95AU.') (**10 points**) 

In [22]:
f = open( "message_to_stations.txt" , 'w' )

if expected_stops >= 1:
    arr_of_stops = np.arange( 1 , expected_stops+1)
    f.write(f"My trip starts at {start_location} and ends at {destination_location}. ")
    
    for i in range(len(arr_of_stops)):
        f.write(f"I will need to encounter a stop at {arr_of_stops[i] * 0.65} AU.\n")
        
if expected_stops < 1:
    f.write("No stops needed for your trip.\n")
    
f = open("message_to_stations.txt" , 'r')
    
output_message = f.read()

print(output_message)

f.close()

My trip starts at Earth and ends at Mars. I will need to encounter a stop at 0.65 AU.
I will need to encounter a stop at 1.3 AU.
I will need to encounter a stop at 1.9500000000000002 AU.



In [28]:
#I will now update my function by copying and pasting the code made here in Problem 4 to my existing planet_tour function.

C) Import the updated function and run it for a trip from Io to Jupiter and Earth to Venus (**5 points**). 

Rename the output message to message_to_stations_1.txt (explain why you should only have 1 message here). Then run it for a trip from the Earth to Pluto (**5 points**). Rename the output message to message_to_stations_2.txt 

In [23]:
from midterm_funct_diegobenavidesucf import planet_tour
help(planet_tour)

Help on function planet_tour in module midterm_funct_diegobenavidesucf:

planet_tour(start, destination)
    This function will take in the given starting location and the given destination location, and return the
    distance between the two locations for the best case and the time taken to travel between them 
    moving at 0.001 AU/hr for the best case.
    The function also returns how many stops of refueling are required, and makes a .txt file to send to 
    the stations mentioning where the refueling stops are needed. It also prints the date for the best case with the 
    distance for the best case of each refuel stop with a 4 digit precision.
    
    INPUT: start location input at the beginning of the code, destination location input in the beginning 
    OUTPUT: distance between the two inputs for the best case, 
            and the travel time between the two places with a speed of 0.001 AU/hr for the best case



In [24]:
#Io to Jupiter case. I will change the output message to message_to_stations_1.txt
#because I dont want to overwrite my message_to_stations.txt

print(planet_tour( 'Io' , 'Jupiter' ))

Type in your starting location and then type in your desired destination location:

Io
Jupiter
Both Io and Jupiter are valid locations. The Journey will proceed!

The best date to travel is Date 3 , the distance is 0.0018 AU.

Your trip will take 0 year(s), 0 day(s), 1 hour(s), and 47 minute(s)

Your journey will require -0.0 refueling stops.
No stops needed for your trip.

('The best date to travel is Date 3 , the distance is 0.0018 AU.', 'Your trip will take 0 year(s), 0 day(s), 1 hour(s), and 47 minute(s)')


In [25]:
#Earth to Venus. I should only have one message here because it will overwrite the meassages with the same title.
#This output will be renamed message_to_stations_2.txt

print(planet_tour( 'Earth' , 'Venus' ))

Type in your starting location and then type in your desired destination location:

Earth
Venus
Both Earth and Venus are valid locations. The Journey will proceed!

The best date to travel is Date 1 , the distance is 0.3915 AU.

Your trip will take 0 year(s), 16 day(s), 7 hour(s), and 30 minute(s)

Your journey will require -0.0 refueling stops.
No stops needed for your trip.

('The best date to travel is Date 1 , the distance is 0.3915 AU.', 'Your trip will take 0 year(s), 16 day(s), 7 hour(s), and 30 minute(s)')


In [26]:
# Earth to Pluto. This output will be renamed message_to_stations_2b.txt for the same reasons stated above

print(planet_tour( 'Earth' , 'Pluto'))

Type in your starting location and then type in your desired destination location:

Earth
Pluto
Both Earth and Pluto are valid locations. The Journey will proceed!

The best date to travel is Date 2 , the distance is 34.4110 AU.

Your trip will take 3 year(s), 338 day(s), 18 hour(s), and 59 minute(s)

Your journey will require 52.0 refueling stops.


My trip starts at Earth and ends at Pluto. I will need to encounter a stop at 0.6500 AU.
I will need to encounter a stop at 1.3000 AU.
I will need to encounter a stop at 1.9500 AU.
I will need to encounter a stop at 2.6000 AU.
I will need to encounter a stop at 3.2500 AU.
I will need to encounter a stop at 3.9000 AU.
I will need to encounter a stop at 4.5500 AU.
I will need to encounter a stop at 5.2000 AU.
I will need to encounter a stop at 5.8500 AU.
I will need to encounter a stop at 6.5000 AU.
I will need to encounter a stop at 7.1500 AU.
I will need to encounter a stop at 7.8000 AU.
I will need to encounter a stop at 8.4500 AU.
I will

In [26]:
#I now reverted my message file back to message_to_stations.txt so that the test cases above dont get overwritten within 
#my midterm folder.

## 5. On the dot. ( 20 points total)

Traveling in the 21st century from A to B with your car, truck or motorcycle was relatively simple. You just needed to know when you have time to go, and you could have many alternatives to the when that would be. Places A and B would always be there....In 2340 though, traveling is a little bit more complex. Locations A and B are on different bodies of the Solar system that orbit the Sun with different speeds. For example, a trip from Mars to the Earth may only be 33,900,000 miles long on day_1, and on day_2 more than 250,000,000 miles long! *A lot* more time and fuel can be needed for your trip if you pick the wrong date to travel. Let's make the needed updates to the function to make the user's trip as efficient as possible.

A) Update the function you created to read in the files for all possible travel dates. The code should then compare the distance (or time) it takes to do the trip at different dates. The code should print a statement letting the user know when it is the best time to travel and only print the distance and refueling information needed for that best-case trip. (**10 points**)

In [27]:
space_distances_date_1 = np.loadtxt('solar_system_date_1.dat', comments = '#')

space_distances_date_2 = np.loadtxt('solar_system_date_2.dat', comments = '#')

space_distances_date_3 = np.loadtxt('solar_system_date_3.dat', comments = '#')

In [28]:
distance_date_1 = space_distances_date_1[ destination_body_i , start_body_i ]

distance_date_2 = space_distances_date_2[ destination_body_i , start_body_i ]

distance_date_3 = space_distances_date_3[ destination_body_i , start_body_i ]


list_of_distances = [ distance_date_1 , distance_date_2 , distance_date_3]


best_date_distance = np.min(list_of_distances)

if distance_date_1 == best_date_distance :
    best_date = 'Date 1'
        
if distance_date_2 == best_date_distance :
    best_date = 'Date 2' 
        
if distance_date_3 == best_date_distance :
    best_date = 'Date 3'
        

print(f'The best date to travel is {best_date} , the distance is {best_date_distance} AU.\n')

The best date to travel is Date 3 , the distance is 2.023 AU.



B) Now that we are ready to do interplanetary trips we also need our mesages to the refueling stations to be more streamlined. Change the function to ensure that the output distances are formated with a 4 digit accuracy (so 0.6500 AU not 0.65 and not 0.65000001) (**5 points**)


In [29]:
#I will round my numbers to 4-digit accuracy using '%.4f' %(x)
#and update all my code printing a digit by adding this function.

# FOR EXAMPLE I will add this structure to my current function

prec_num = 1.33

print(f"I will need to encounter a stop at %.4f AU.\n" % (prec_num))
        
    
#Now I will update my function with this syntax

I will need to encounter a stop at 1.3300 AU.



C) Test the changes you made for a trip from Earth to Neptune. Rename the output message to message_to_stations_3.txt and save it to your handin folder. (**5 points**)

In [30]:
from midterm_funct_diegobenavidesucf import planet_tour
print(help(planet_tour))

Help on function planet_tour in module midterm_funct_diegobenavidesucf:

planet_tour(start, destination)
    This function will take in the given starting location and the given destination location, and return the
    distance between the two locations for the best case and the time taken to travel between them 
    moving at 0.001 AU/hr for the best case.
    The function also returns how many stops of refueling are required, and makes a .txt file to send to 
    the stations mentioning where the refueling stops are needed. It also prints the date for the best case with the 
    distance for the best case of each refuel stop with a 4 digit precision.
    
    INPUT: start location input at the beginning of the code, destination location input in the beginning 
    OUTPUT: distance between the two inputs for the best case, 
            and the travel time between the two places with a speed of 0.001 AU/hr for the best case

None


In [31]:
print(planet_tour('Earth' , 'Neptune'))

Type in your starting location and then type in your desired destination location:

Earth
Neptune
Both Earth and Neptune are valid locations. The Journey will proceed!

The best date to travel is Date 3 , the distance is 28.8610 AU.

Your trip will take 3 year(s), 107 day(s), 13 hour(s), and 0 minute(s)

Your journey will require 44.0 refueling stops.


My trip starts at Earth and ends at Neptune. I will need to encounter a stop at 0.6500 AU.
I will need to encounter a stop at 1.3000 AU.
I will need to encounter a stop at 1.9500 AU.
I will need to encounter a stop at 2.6000 AU.
I will need to encounter a stop at 3.2500 AU.
I will need to encounter a stop at 3.9000 AU.
I will need to encounter a stop at 4.5500 AU.
I will need to encounter a stop at 5.2000 AU.
I will need to encounter a stop at 5.8500 AU.
I will need to encounter a stop at 6.5000 AU.
I will need to encounter a stop at 7.1500 AU.
I will need to encounter a stop at 7.8000 AU.
I will need to encounter a stop at 8.4500 AU.
I

In [3]:
#I will revert the .txt message to message_to_stations.txt so I dont overwrite the message I just made.

## 6. The grand tour ( 35 points total)

A grand tour wouldn't be a *grand* tour without a stop or two on the way. As planets rotate around the Sun, the 'best' stop might not always be trivial (e.g., going from Ceres to Venus sometimes the best stop is Mars, sometimes it can be Earth, as Mars (or Earth) is in on the opposite side of their orbit). In this last task you will make the necessary changes to the code to let the user do a grand tour with one stop along the way.

In [20]:
from midterm_funct_diegobenavidesucf import planet_tour

A) Update the function to allow for one stop along the way. To do this, the function will get the start and end point of the trip and ask the user if they want to do a tour (input should be Y/N or Yes/No). (**2 points**)


A1) If the user replies 'N' or 'No' the function should work as above and print the best date to travel from A to B etc. (**5 points**)



A2) If the user replies 'Yes' or 'Y', the function will then check for all dates what a possible stop would be. (**13 points**) 

A stop that makes sense needs to have the smallest possible extra distance in comparison to the direct trip from A to B. E.g., going from Mars to Ceres on date 1 the stop the traveler would do is on Mercury adding 'just' 1.3 AU to the trip (because Venus and the Earth are on the 'wrong side' of their orbits). 

The function will then print out the information for the user to see:

    1) on what date can they do, what stop, and 

    2) at what cost in milage and 

    3) number of refueling stops. 
    
Don't save a message for the refueling stations for each of these stops. We will adapt this part later on.

B) Test the updates you did for a trip from Mars to Ceres. (5 points)

C) The code should now let the user choose which is their prefered stop. 

After getting the available stops per date, store them in a dictionary named stops_for_date. The keys should be the name of the stop (ordered for date 1, 2 and 3: e.g., 'Mercury' 'Mercury' 'Venus'), and the values should be the corresponding total distance of the trip (e.g., 4.2058, 4.5132, 4.5077 ). If more than one dates return the same stop make sure to adapt the key accordingly ('Mercury_1', 'Mercury_2' ). Use an If/else if/else structure to do this. **(6 points)**

What would happen if you fogot to do this name change? Explain with a comment in your code. **(2 points)**

The code will then ask the user which is their prefered stop. If more than one dates return the same stop then the code will let the user know when they should travel (the date with the least travel time/miles). **(5 point)** 
<br>(in the above example if the user wanted to stop at Mercury we would let them know the best time is date 1).
    Tip: to do this you can use an if/else if/ else structure to test if more than one dates match the user's input and if so decide when they should move. You can also *try* different methods.

The code will then save the necessary message for the refueling stations on that date. **( 2 points)**

The message should now let the stations know on what date they will meet! The message should print the start date of the trip (date 1, 2, 3) on the start of the message. 


(e.g., :

############ refueling information##############
############

########## Starting trip from Mars to Ceres on date 1 #########

######### Will refuel on date 1 + 3.01 days at location 0.6500 AU

######### Will refuel on date 1 + 5.52 days at location 1.3000 AU

Note that this is a random example, not based on any of the data)

## 7. The wraping up. (10 points)


Prepare and submit your homework. Copy the finalized Jupyter notebook to the handin/midterm_* folder and don’t forget to commit and push it to GitHub. Explain what you did to do that in your log. Make a screenshot that shows you committed the file and add it to your handin/midterm_* folder (remember to use an appropriate name for the screenshot!). Make sure that *all* necessary files, including data files are included in the folder. Write what you did to make and submit the zip file into your log. When satisfied, close the log, copy it to your homework directory one last time, and run the commands to make and submit the zip file. Turn the file in on WebCourses. 

In [21]:
#I will know copy and paste my log int my midterm_diegobenavidesucf folder, push origin from github desktop app,
#and screenshot the proof of pushing origin
#I will then zip the file and submit